launchSearch.ts

```diff
@@ -0,0 +1,73 @@
+/*!
+ * Copyright © 2023 United States Government as represented by the
+ * Administrator of the National Aeronautics and Space Administration.
+ * All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import Dockerode from 'dockerode'
+
+type Message = {
+  action: string
+}
+
+const [, , argv] = process.argv
+const { dataDir, logsDir, engine, port, options } = JSON.parse(argv)
+
+let dockerContainer: Dockerode.Container
+
+async function launchDocker() {
+  const Image =
+    engine === 'elasticsearch'
+      ? 'elastic/elasticsearch:8.6.2'
+      : 'opensearchproject/opensearch:2.11.0'
+  console.log('Launching Docker container', Image)
+  const docker = new Dockerode()
+
+  const container = await docker.createContainer({
+    Env: [...options, 'path.data=/var/lib/search', 'path.logs=/var/log/search'],
+    HostConfig: {
+      AutoRemove: true,
+      Mounts: [
+        { Source: dataDir, Target: '/var/lib/search', Type: 'bind' },
+        { Source: logsDir, Target: '/var/log/search', Type: 'bind' },
+      ],
+      PortBindings: {
+        [`${port}/tcp`]: [{ HostIP: '127.0.0.1', HostPort: `${port}` }],
+      },
+    },
+    Image,
+  })
+  const stream = await container.attach({ stream: true, stderr: true })
+  stream.pipe(process.stderr)
+  await container.start()
+  return container
+}
+
+async function waiting() {
+  await dockerContainer.wait()
+  process.exit(0)
+}
+
+async function launch() {
+  dockerContainer = await launchDocker()
+  waiting()
+}
+
+process.once('message', async (message: Message) => {
+  if (message.action === 'kill') {
+    await dockerContainer.kill()
+    process.exit(0)
+  }
+})
+
+launch()
+
+const signals = ['SIGTERM', 'SIGINT']
+signals.forEach((signal) => {
+  process.on(signal, async () => {
+    await dockerContainer.kill()
+    process.exit(0)
+  })
+})
```

Comment:
- You might not need the `process.exit` call because Node.js will exit once the run loop has no more promises to settle.
- You can combine all there event handlers like this:

```suggestion
const dockerContainer = await launchDocker()

['message', 'SIGTERM', 'SIGINT'].forEach((signal) => {
  process.on(signal, async () => {
    await dockerContainer.kill()
    process.exit(0)
  })
})

await dockerContainer.wait()
```

Reply:
Can we change the target? Since this is just an Architect plugin it's not like we'll ever bundle it into a web app.